### PR TITLE
Update flash_attention to version 2.0.1

### LINKED ISSUE #4322: [flashattn] Support Flash Attention to v2.0.1
According [flash-attention]( https://github.com/Dao-AILab/flash-attention), I modify the code for supporting flashattn v2.0.1

### PR DESCRIPTION
**Pull request by kkk55596**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
> 
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

fixed #4322 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

According to [Flash Attention](https://github.com/Dao-AILab/flash-attention), I modify `flash_attention.py` and `requirements-test.txt` for supporting **flashattn v2.0.1**

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by clyang** [comment]
wow. According to Flash Attention's webiste
> FlashAttention-2 is about 2x faster than its previous version, reaching up to 230 TFLOPs/s on A100 GPUs (FP16/BF16).

It would be great if this PR can pass the review and be accepted.

**Comment by kurisusnowdeng** [comment]
@kkk55596 Thank you so much for contribution. We will actually provide the general `ColoAttention` interface and deprecate the other ones very soon. Thus, could you please try to replace the xformers part that we currently use with flash attention 2?

Here are few tips for you:

1. We could use `flash_attn_func` for attention with no paddings
2. Attention with paddings will need `get_seq_info_from_mask`, `unpad`, `repad` in order to work with `flash_attn_varlen_func`
3. Flash attention 2 only supports fp16/bf16 on Ampere or better GPUs. For other precisions or hardwares, we still need xformers to accelerate attention;
4. (Optional) flash attention's CUDA version does not support attention bias, while its triton version does. We would really appreciate it if you are able to help `ColoAttention` to support attention bias.

File: colossalai/kernel/cuda_native/flash_attention.py [hunk 1]
```diff
@@ -190,9 +190,9 @@ def triton_cuda_check():
 try:
     from flash_attn.flash_attention import FlashAttention
     from flash_attn.flash_attn_interface import (
-        flash_attn_unpadded_func,
-        flash_attn_unpadded_kvpacked_func,
-        flash_attn_unpadded_qkvpacked_func,
+        flash_attn_varlen_func,
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/flash_attention.py:193]: As commented in conversations

**Review by kurisusnowdeng — CHANGES_REQUESTED** [review]

**Comment by kurisusnowdeng** [comment]
Closed since fixed in #4347

**Pull request closed without merging**